# ``Statistics Exercise`` - **``Solution``**

## __Import Library__

In [1]:
import pandas as pd
pd.set_option('display.max_columns', None)

## __Open Dataset__

In [2]:
df = pd.read_csv('melb_data.csv')
df.head()

Suburb           Address  Rooms Type      Price Method SellerG  \
0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  Postcode  Bedroom2  Bathroom  Car  Landsize  \
0  3/12/2016       2.5    3067.0       2.0       1.0  1.0     202.0   
1  4/02/2016       2.5    3067.0       2.0       1.0  0.0     156.0   
2  4/03/2017       2.5    3067.0       3.0       2.0  0.0     134.0   
3  4/03/2017       2.5    3067.0       3.0       2.0  1.0      94.0   
4  4/06/2016       2.5    3067.0       3.0       1.0  2.0     120.0   

   BuildingArea  YearBuilt CouncilArea  Lattitude  Longtitude  \
0           NaN        NaN       Yarra   -37.7996    144.9984   
1          79.0     1900.0       Yarra   -37.8079    144.9934   
2         150.0     1900.0       Yarra   -37.8093    144.9944   
3           NaN        NaN       Yarra   -37.7969    144.9969   
4         142.0     2014.0       Yarra   -37.8072    144.9941   

              Regionname  Propertycount  
0  Northern Metropolitan         4019.0  
1  Northern Metropolitan         4019.0  
2  Northern Metropolitan         4019.0  
3  Northern Metropolitan         4019.0  
4  Northern Metropolitan         4019.0

<hr>

## **Nomor 1**

Tampilkan statistik deskriptif pada variabel **Rooms, Distance, Landsize**, dan **Price**!

In [3]:
df[['Rooms', 'Distance', 'Landsize', 'Price']].describe()

Rooms      Distance       Landsize         Price
count  13580.000000  13580.000000   13580.000000  1.358000e+04
mean       2.937997     10.137776     558.416127  1.075684e+06
std        0.955748      5.868725    3990.669241  6.393107e+05
min        1.000000      0.000000       0.000000  8.500000e+04
25%        2.000000      6.100000     177.000000  6.500000e+05
50%        3.000000      9.200000     440.000000  9.030000e+05
75%        3.000000     13.000000     651.000000  1.330000e+06
max       10.000000     48.100000  433014.000000  9.000000e+06

<hr>

## **Nomor 2**

a) Berapa jumlah **Rooms** pada rumah yang harganya paling mahal?


In [4]:
df[df['Price'] == df['Price'].max()][['Rooms']]

Rooms
12094      3

b) Berapa rata-rata harga rumah yang jarak ke CBD (variabel **Distance**) terdekat?


In [5]:
df[df['Distance'] == df['Distance'].min()]['Price'].mean()

671166.6666666666

c) Berapa rata-rata harga rumah yang luas tanahnya (variabel **Landsize**) 1 kali standar deviasi di atas rata-rata luas tanah?


In [6]:
land_mean = df['Landsize'].mean()
land_std = df['Landsize'].std()

df[df['Landsize'] > (land_mean + land_std)]['Price'].mean()

880776.6666666666

d) Berapa rentang (_range_) harga rumah (variabel **Price**) pada region ``Southern Metropolitan ``?

In [7]:
df_sm = df[df['Regionname'] == 'Southern Metropolitan']

price_range = df_sm['Price'].max() - df_sm['Price'].min()
price_range

7869000.0

<hr>

## **Nomor 3**

* Nilai outlier adalah nilai yang berada di bawah ``Q1 - (1.5 * IQR)`` atau di atas ``Q3 + (1.5 * IQR)``
* Hitung dulu Q1, Q2, dan IQR pada pada kolom **Price** (harga rumah)
* Hitung ada berapa nilai outlier pada kolom **Price**!

In [8]:
import numpy as np

df['Price'] = df['Price'].apply(np.int64)

In [9]:
q1 = df['Price'].describe()['25%']
q3 = df['Price'].describe()['75%']

iqr = q3 - q1
iqr

680000.0

In [10]:
batas_bawah = int(q1 - (1.5 * iqr))
batas_atas = int(q3 + (1.5 * iqr))

In [11]:
outlier_price = df[(df['Price'] < batas_bawah) | (df['Price'] > batas_atas)][['Price']]
outlier_price.head()

Price
80   2850000
92   2615000
97   2575000
102  3010000
103  2800000

In [12]:
# jumlah outlier pada kolom 'Price'
len(outlier_price)

612

In [13]:
outlier_price.index

Int64Index([   80,    92,    97,   102,   103,   108,   112,   118,   124,
              233,
            ...
            13466, 13468, 13474, 13483, 13518, 13521, 13523, 13553, 13555,
            13578],
           dtype='int64', length=612)

<hr>

## **Nomor 4**

#### **``Shapiro Wilk Test``**

Menguji asumsi normalitas variabel ``Price``

#### **Interpretation**
* If the P-Value of the Shapiro Wilk Test is larger than 0.05, we assume a normal distribution
* If the P-Value of the Shapiro Wilk Test is smaller than 0.05, we do not assume a normal distribution

In [14]:
from scipy.stats import shapiro

sw_statistic, sw_pvalue = shapiro(df['Price'])

if sw_pvalue > 0.05:
    print(f'P-value: {sw_pvalue}. So, we assume a normal distribution')
else:
    print(f'P-value: {sw_pvalue}. So, we do not assume a normal distribution')

P-value: 0.0. So, we do not assume a normal distribution


c:\users\moham\appdata\local\programs\python\python38\lib\site-packages\scipy\stats\morestats.py:1681: UserWarning: p-value may not be accurate for N > 5000.
  warnings.warn("p-value may not be accurate for N > 5000.")


#### **``D’Agostino and Pearson’s Test``**

Menguji asumsi normalitas variabel ``Price``

#### **Interpretation**
* If the P-Value of the D'Agostino-Pearson Test is larger than 0.05, we assume a normal distribution
* If the P-Value of the D'Agostino-Pearson Test is smaller than 0.05, we do not assume a normal distribution

In [15]:
from scipy.stats import normaltest
dp_statistic, dp_pvalue = normaltest(df['Price'])

if dp_pvalue > 0.05:
    print(f'P-value: {dp_pvalue}. So, we assume a normal distribution')
else:
    print(f'P-value: {dp_pvalue}. So, we do not assume a normal distribution')

P-value: 0.0. So, we do not assume a normal distribution


<hr>

## **Nomor 5**

Jika diasumsikan telah memenuhi syarat statistik parametrik, hitung seberapa kuat korelasi variabel **Rooms** dengan **Price**, serta arah korelasinya!

In [16]:
df['Rooms'].corr(df['Price'], method = 'pearson')

0.49663367618652987

<hr>

## **Nomor 6**

Jika diasumsikan telah memenuhi syarat statistik parametrik, dari beberapa Regionname manakah **Regionname** yang variabel jumlah kamarnya (**Rooms**) memiliki korelasi terkuat dengan variabel harga rumah (**Price**)!

In [17]:
df['Regionname'].unique()

array(['Northern Metropolitan', 'Western Metropolitan',
       'Southern Metropolitan', 'Eastern Metropolitan',
       'South-Eastern Metropolitan', 'Eastern Victoria',
       'Northern Victoria', 'Western Victoria'], dtype=object)

In [18]:
for i in df['Regionname'].unique():
    print(f'{i} = ' + str(df[df['Regionname'] == i]['Rooms'].corr(df[df['Regionname']==i]['Price'], method= 'pearson')))

Northern Metropolitan = 0.39281345830224074
Western Metropolitan = 0.47603029384349793
Southern Metropolitan = 0.6837438462796437
Eastern Metropolitan = 0.5029928557468845
South-Eastern Metropolitan = 0.2677994408114044
Eastern Victoria = 0.31442339534695335
Northern Victoria = 0.36114429890812727
Western Victoria = 0.5518865952271457


In [19]:
corr = []
for i in df['Regionname'].unique():
    corr.append(df[df['Regionname'] == i]['Rooms'].corr(df[df['Regionname']==i]['Price'], method= 'pearson'))
    
pd.DataFrame(data = corr, index = df['Regionname'].unique(), columns = ['Correlation Price vs Rooms']
            ).sort_values('Correlation Price vs Rooms', ascending=False)

Correlation Price vs Rooms
Southern Metropolitan                         0.683744
Western Victoria                              0.551887
Eastern Metropolitan                          0.502993
Western Metropolitan                          0.476030
Northern Metropolitan                         0.392813
Northern Victoria                             0.361144
Eastern Victoria                              0.314423
South-Eastern Metropolitan                    0.267799

<hr>

## **Nomor 7**

Jika diasumsikan **tidak** memenuhi syarat statistik parametrik, hitung seberapa kuat korelasi variabel **Distance** dengan **Price**, serta arah korelasinya!

In [20]:
df['Distance'].corr(df['Price'], method='spearman')

-0.12998984131880323

<hr>

## **Nomor 8**

Jika diasumsikan telah memenuhi syarat statistik parametrik, dari beberapa Regionname manakah **Regionname** yang variabel jumlah jarak ke CBD (**Distance**) paling tidak berkorelasi dengan variabel harga rumah (**Price**)! Serta tampilkan juga Regionname yang paling berkorelasi kedua variabel ini!

In [21]:
for i in df['Regionname'].unique():
    print(f'{i} = ' + str(df[df['Regionname'] == i]['Distance'].corr(df[df['Regionname']==i]['Price'], method= 'pearson')))

Northern Metropolitan = -0.368551996179051
Western Metropolitan = -0.33408555047092137
Southern Metropolitan = 0.002489149218933464
Eastern Metropolitan = -0.12435562852683693
South-Eastern Metropolitan = -0.2855640518908052
Eastern Victoria = -0.34569916122464606
Northern Victoria = 0.08956649327360579
Western Victoria = -0.04606415345276126


In [22]:
corr = []
for i in df['Regionname'].unique():
    corr.append(df[df['Regionname'] == i]['Distance'].corr(df[df['Regionname']==i]['Price'], method= 'pearson'))
    
pd.DataFrame(data = corr, index = df['Regionname'].unique(), columns = ['Correlation Price vs Distance']
            ).sort_values('Correlation Price vs Distance', ascending=True)

Correlation Price vs Distance
Northern Metropolitan                           -0.368552
Eastern Victoria                                -0.345699
Western Metropolitan                            -0.334086
South-Eastern Metropolitan                      -0.285564
Eastern Metropolitan                            -0.124356
Western Victoria                                -0.046064
Southern Metropolitan                            0.002489
Northern Victoria                                0.089566

* Regionname yang paling tidak berkorelasi atau korelasinya paling lemah di dua variabel di atas adalah **Sourthern Metropolitan** (0.002489 atau 0.24 persen).
* Regionname yang paling berkorelasi di dua variabel di atas adalah **Northern Metropolitan** (-0.368552 atau - 36,86 persen)

<hr>

## **Nomor 9**

Uji hipotesis apakah metode pembelian/**Method** berkaitan dengan **Type** / tipe rumah (dependen) atau tidak saling berkaitan (independen)!

In [23]:
df_ct = pd.crosstab(df['Type'], df['Method'])
df_ct

Method    PI     S  SA    SP   VB
Type                             
h       1069  6507  66  1079  728
t        134   723   7   143  107
u        361  1792  19   481  364

#### **Hypothesis:**
* Ho : The two variables are independent
* Ha :  The two variables are dependent.

In [24]:
from scipy.stats import chi2_contingency

stat, p, dof, expected = chi2_contingency(df_ct)

# conclusion
alpha = 0.05
print('P value:', p)

if p <= alpha:
    print('Dependent (tolak H0)')
else:
    print('Independent (terima H0)')

P value: 3.183702485154384e-22
Dependent (tolak H0)


<hr>

## **Nomor 10**

Uji hipotesis apakah rata-rata harga rumah (**Price**) antar Regionname berbeda ataukah sama!

Diasumsikan persyaratan statistik parametrik terpenuhi

**Hipotesis**:
* Ho : μ1 = μ2 = … = μk (Rata-rata harga rumah antar Regionname sama)  
* Ha : Setidaknya ada satu pasang Regionname yang rata-rata harga rumahnya tidak sama, tapi kita tidak tahu yang mana

In [25]:
df['Regionname'].unique()

array(['Northern Metropolitan', 'Western Metropolitan',
       'Southern Metropolitan', 'Eastern Metropolitan',
       'South-Eastern Metropolitan', 'Eastern Victoria',
       'Northern Victoria', 'Western Victoria'], dtype=object)

In [26]:
from scipy.stats import f_oneway

f_oneway(df[df['Regionname'] == 'Northern Metropolitan']['Price'],
         df[df['Regionname'] == 'Western Metropolitan']['Price'],
         df[df['Regionname'] == 'Southern Metropolitan']['Price'],
         df[df['Regionname'] == 'Eastern Metropolitan']['Price'],
         df[df['Regionname'] == 'South-Eastern Metropolitan']['Price'],
         df[df['Regionname'] == 'Eastern Victoria']['Price'],
         df[df['Regionname'] == 'Northern Victoria']['Price'],
         df[df['Regionname'] == 'Western Victoria']['Price'],
        )

F_onewayResult(statistic=284.40018651008097, pvalue=0.0)

Jika p-value kurang dari sama dengan 0.05 maka cukup bukti untuk menolak Ho.
Berarti Ha diterima ada pasangan Regionname yang perbedaan rata-rata harga rumahnya signifikan, tapi tidak tahu yang mana.

In [27]:
values = []
for i in df['Regionname'].unique():
    values.append(round(df[df['Regionname']==i]['Price'].mean(), 2))
    
pd.DataFrame(data=values, index=df['Regionname'].unique(), columns=['Price Mean']
            ).sort_values(by='Price Mean', ascending=False)

Price Mean
Southern Metropolitan       1372963.37
Eastern Metropolitan        1104079.63
South-Eastern Metropolitan   922943.78
Northern Metropolitan        898171.08
Western Metropolitan         866420.52
Eastern Victoria             699980.79
Northern Victoria            594829.27
Western Victoria             397523.44

<hr>

## **Thank you & Keep Learning**